<div>
    <table>
        <tr>
            <td>
                    <h1>Brightway2 seminar</h1>
                    Chris Mutel <a href="https://www.psi.ch/">(PSI)</a>
                    Pascal Lesage <a href="http://www.ciraig.org/en/">(CIRAIG)</a>
            </td>
        </tr>
    </div>

# Brightway modules: bw2io, bw2data, bw2calc

*The following content is taken and adapted from [the course materials dispensed at the 2022 Brightcon conference](https://github.com/Depart-de-Sentier/teaching-material/tree/main/beginners/Projects%2C%20databases%2C%20exchanges%2C%20activities). If you need additional notebooks, go to this repository*.


<div class="alert alert-info">
Note: we will be using Brightway2, not Brightway25. From the user end side, very little differ between the two.
</div>





**Brightway2 documentation** 👉 [https://docs.brightway.dev/en/legacy/index.html](https://docs.brightway.dev/en/legacy/index.html)

**Brightway25 documentation** 👉 [https://docs.brightway.dev/en/latest/index.html](https://docs.brightway.dev/en/latest/index.html)

### Learning objectives  
  - Learn about the general structure of Brightway and its most importand abstractions: projects, databases, activities and exchanges  
  - Learn how to find objects (notably activities and exchanges), assign them to variables and work with them using their associated methods  
  - Learn about how simple LCA calculations are done (one product, one impact category), and specifically how the different matrices are built and used  
  - Learn how to extract information from the matrices (inputs or results) and translate them into nice, human-readable objects  
  - Learn different ways to carry out comparative LCAs  
  - Learn different ways to carry out LCAs with multiple impact categories

### Content  

#### 1) Getting started  
##### 1.1) Jupyter lab / notebooks and Brightway2  
  - A quick tour of jupyter lab / notebooks  
  - Accessing Brightway2 libraries  
  
##### 1.2) Projects  
  - Concept  
  - Creating a new project, or switching to an existing project  
  - Contents of a project  

  
##### 1.3) bw2_setup(), biosphere3 database and LCIA methods  
  - bw2_setup()  
  - biosphere3 database and a first look at database objects  
  - Getting activities from codes or keys  
  - Methods  
  - Looking up elementary flows (list comprehensions, search)  
  - Searching for methods  
  - Nice display of data in methods 

##### 1.4) LCI databases  
  - Importing (succinct)  
  - LCI activities  
  - Looking up activities  
  - LCI exchanges
  - Loaded LCI databases
  
#### 2) My first LCA - simplest case:  
##### 2.1) General syntax of LCA calculations  

##### 2.2) The `demand` attribute  

##### 2.3) Reminder of the system that needs to be solved in calculating an LCI  

##### 2.4) Building the matrices  

  - $A$ matrix  
  - $B$ matrix  
  - $f$ (demand array)  
  
##### 2.5) Solution to the inventory calculation  

  - Supply array  
  - Inventory matrix  
  
##### 2.6) Life Cycle Impact Assessment  

  - `.lcia()` method  
  - Simple contribution analysis  
  
#### 3) My second LCA - comparative LCA:
    
#### 4) My third LCA - Multiple impact categories
  
#### 5) My first and third LCAs revisited with MultiLCA

### 1) Getting started

#### 1.1) Jupyter lab / notebooks and accessing Brightway2

This teaching material was produced with the intention to be used in Jupyter Notebooks.

For installation instructions, consult the [brightway documentation](https://docs.brightway.dev/en/legacy/content/installation/installation.html)

<div class="alert alert-info">
This notebook is meant to run with a specific conda (and python) environment.
    Make sure you have started Jupyter Notebooks or Jupyter Lab from the conda
    environment where the brightway2 modules are installed.
</div>

##### Accessing Brightway2 libraries

The different modules in Brightway2 are Python libraries. This means that, to use them, you can use any environment from which you normally use Python (Idle, command prompt, Spyder or, as is the case today, Jupyter Notebooks).  

We will favour Jupyter Notebooks in this course because it allows us to integrate code and text. It will also allow us to provide code snippets for you to complete.  

Like all other Python packages, you need to `import` Brightway2 modules. We will here import it as `bw`. This means that, to access a **method** or **function** from the brightway2 modules, the prefix `bw.` needs to be added. 

In [1]:
import brightway2 as bw

Alternatively, we can import the Brightway's sub-modules directly:

In [ ]:
import bw2io, bw2data, bw2calc

We're also going to be using the following libraries:

In [2]:
import os               # to use "operating system dependent functionality"
import numpy as np      # "the fundamental package for scientific computing with Python"
import pandas as pd     # "high-performance, easy-to-use data structures and data analysis tools" for Python

#### 1.2) Projects

##### Concept

The top-level containent in Brightway2 is the project (see [here](https://2.docs.brightway.dev/intro.html#projects) for a description and [here](https://2.docs.brightway.dev/technical/bw2data.html#projects) for the **technical docs about projects**). A project contains LCI databases, LCIA methods and other less often used objects. Objects from one project do not interract with objects within other projects. By analogy, projects are like databases in other LCA software.  
When you first launch Brightway2, you will be in the `default` project. You can check this using the `current` property of the `projects` object: 

In [8]:
bw.projects

Brightway2 projects manager with 16 objects:
	arezoo
	bw25
	coursePSI
	default
	ei38
	ei39
	new
	new4
	polyviz
	psi1
	strat_metals
	test
	testbio
	testbio2
	testbio3
	uvek
Use `projects.report()` to get a report on all projects.

In [7]:
bw.projects.set_current("coursePSI")

##### Creating a new project, or switching to an existing project

Let's create a new project for this seminar, unsurprisingly called "bw2_seminar_2017". There are two ways of doing this:  
* `projects.create_project('brightcon22')` will create the project, but you will remain in your current project.
* `projects.set_current('brightcon22')` will switch you to the project passed as argument, and create it first if it doesn't exist.  Let's do the latter:

In [ ]:
# The name of the project is entered as string; 
# it doesn't really have any restrictions, so can include spaces, 
# special characters, other languages, or even emoji.



You can always see what projects you have on your computer by running `list(bw.projects)`. Unless you have worked with Brightway2 before on your computer, your list should contain two projects: 'default' and 'bw2_seminar_2017'.

_**Exercise**_: list the projects on your computer.

In [14]:
bw.projects

Brightway2 projects manager with 16 objects:
	arezoo
	bw25
	coursePSI
	default
	ei38
	ei39
	new
	new4
	polyviz
	psi1
	strat_metals
	test
	testbio
	testbio2
	testbio3
	uvek
Use `projects.report()` to get a report on all projects.

Like in all Python modules, you can get additionnal information on the `projects` object and associated properties and methods by typing `help(projects)`. The [docs](https://docs.brightwaylca.org/technical/bw2data.html#projects) are of course more verbose.

##### Contents of a project

One property of `projects` is its location, given by `projects.dir`:

In [18]:
bw.projects.dir

'/Users/romain/Library/Application Support/Brightway3/coursePSI.ad34cb91313c17c13fa891538321d355'

<div class="alert alert-info">
    Try to locate your project directory on your computer.
</div>

First things first: **do not panic**! You can use Brightway2 for years without ever opening this directory, but we will discuss some of these files later.

It is simply interesting to note that, for now, all the directories are empty except the `lci` directory, which contains an empty database.

All in all, the project takes up about 150KB.  
Its now time to start populating the project.

#### 1.3) bw2setup(), biosphere3 database and LCIA methods

##### bw2setup()

The first thing you should do is add flow and LCIA methods. This is done by running the `bw2setup` function:

In [20]:
bw.bw2setup()

Writing activities to SQLite3 database:


Creating default biosphere

Applying strategy: normalize_units
Applying strategy: drop_unspecified_subcategories
Applying strategy: ensure_categories_are_tuples
Applied 3 strategies in 0.00 seconds


0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:00


Title: Writing activities to SQLite3 database:
  Started: 03/19/2024 10:09:25
  Finished: 03/19/2024 10:09:26
  Total time elapsed: 00:00:00
  CPU %: 73.00
  Memory %: 2.90
Created database: biosphere3
Creating default LCIA methods

Wrote 762 LCIA methods with 227223 characterization factors
Creating core data migrations



In [21]:
bw.databases

Databases dictionary with 1 object(s):
	biosphere3

The output tells us that bw2_setup created some very useful things:  
  - Created a database called "biosphere3": this database contains elementary flows (called biosphere exchanges in Brightway2)  
  - **762** impact assessment methods  (with bw2data version 3.6.6)

<div class="alert alert-info">
    Note: The biosphere database contains all "natural" flows human-made activities connect to. E.g., CO2 emissions, ore in ground, etc.
</div>
  
It also created some a `mapping` between the imported exchanges and some integer: more on this later.  
The whole directory now takes up 157MB.

##### biosphere3 database and a first look at database objects

<div class="alert alert-info">
    Note: You can always list the databases inside a project by simply typing 'bw.databases'. This accesses the 'database.json' file in your 'project.dir' (I learned the latter by typing `databases?`, you should try it too.)
</div>



In [22]:
bio = bw.Database("biosphere3")

In [23]:
bio

Brightway2 SQLiteBackend: biosphere3

While not impossible to interact with the data at this level, you probably never will unless you are developping some funky program. Instead, it is strongly recommended to learn to work with `abstractions`.  

To access a database in Brightway, you use the `Database` initialization method (again, you can type `Database?` for more information - this is the last time I'll mention this.

It doesn't actually return anything other than information about the Backend.  
However, there are many properties and functions associated with this database object.  These are found [here](https://2.docs.brightway.dev/technical/bw2data.html#databasechooser). We can also have a look through the autocomplete. Let's assign the database to a variable:

Let's check the my_bio `type`:

In [24]:
type(bio)

bw2data.backends.peewee.database.SQLiteBackend

Let's check its length:

In [32]:
len(bio)

4709

This is exactly the number of items we saw had been added to databases.db  
Given this, what do you think is going on?

If you type `my_bio.` and click on tab, you should get a list of properties and methods associated with database objects. Try this now:

In [38]:
bio.random(filters=True, true_random=True)

'Acetic acid, trifluoro-' (kilogram, None, ('air',))

Some of the more basic ones we will be using now are :  
  - `random()` - returns a random activity in the database
  - `get(*valid_exchange_tuple*)` - returns an activity, but you must know the activity key
  - `load()` - loads the whole database as a dictionary.
  - `make_searchable` - allows searching of the database (by default, it is already searchable)
  - `search` - search the database  
  
Lets start with `random`:

In [49]:
random_bio = bio.random()
random_bio

'Zinc-65' (kilo Becquerel, None, ('air',))

This returns a biosphere activity, but without assigning it to a variable, there is not much we can do with it directly.  

Note: It may seem counter-intuitive for elementary *flows* to be considered *activities* in Brightway, but it is no mistake. 
LCA models are made up of **nodes** (activities) that are linked by **edges** (exchanges). The biosphere activities are the nodes *outside* the technosphere. Emissions and resource extractions are modelled as exchanges between activities in the technosphere (part of the product system) and these biosphere activities.  

More on this later. 

For now, let's assign another random activity to a variable:

In [50]:
random_bio.as_dict()

{'categories': ('air',),
 'code': '4aed2261-0461-4014-be23-cf83de76a3b4',
 'CAS number': '013982-39-3',
 'name': 'Zinc-65',
 'database': 'biosphere3',
 'unit': 'kilo Becquerel',
 'type': 'emission'}

We can get the type of the object that was returned from the database:

In [52]:
type(random_bio)

bw2data.backends.peewee.proxies.Activity

The type is an **activity proxy**. Activity proxies allow us to interact with the content of the database.

In Brightway, we *almost* always work with `Activity` or `Exchange` objects. 

To see what the activity contains, we can convert it to a dictionary:

In [ ]:
random_bio.as_dict()

##### Getting activities from codes or keys

We can see that the activities in the biosphere3 database have unique **codes**, which we can use with the `get` function:

In [42]:
bio.get(random_bio["code"])

'Plutonium-238' (kilo Becquerel, None, ('air', 'low population density, long-term'))

Activities can also be "gotten" via `get_activity`, but the argument is the activity **key**, consisting of a tuple with two elements: the database name, and the activity code.

**Exercise:** Use `bw.get_activity()` to retrieve the random biosphere activity. 

In [47]:
random_bio["code"]

'3ae6e5bb-fdd6-424a-be6f-80e7240ad8b7'

In [55]:
bw.get_activity(random_bio.key)

'Zinc-65' (kilo Becquerel, None, ('air',))

You can always find (or return) the key to an activity using the `.key` property.

In [54]:
random_bio.key

('biosphere3', '4aed2261-0461-4014-be23-cf83de76a3b4')

##### Searching for activities

Let's say we are looking for a specific elementary flow, we can use the `search` method of the database (see [here](https://docs.brightwaylca.org/technical/bw2data.html#default-backend-databases-stored-in-a-sqlite-database) for more details on using search):

In [56]:
bio.search("carbon dioxide")

['Carbon dioxide, fossil' (kilogram, None, ('air', 'urban air close to ground')),
 'Carbon dioxide, fossil' (kilogram, None, ('air', 'low population density, long-term')),
 'Carbon dioxide, fossil' (kilogram, None, ('air',)),
 'Carbon dioxide, fossil' (kilogram, None, ('air', 'non-urban air or from high stacks')),
 'Carbon dioxide, in air' (kilogram, None, ('natural resource', 'in air')),
 'Carbon dioxide, fossil' (kilogram, None, ('air', 'lower stratosphere + upper troposphere')),
 'Carbon dioxide, non-fossil' (kilogram, None, ('air', 'low population density, long-term')),
 'Carbon dioxide, non-fossil' (kilogram, None, ('air', 'urban air close to ground')),
 'Carbon dioxide, non-fossil' (kilogram, None, ('air',)),
 'Carbon dioxide, non-fossil' (kilogram, None, ('air', 'lower stratosphere + upper troposphere')),
 'Carbon dioxide, non-fossil' (kilogram, None, ('air', 'non-urban air or from high stacks')),
 'Carbon dioxide, from soil or biomass stock' (kilogram, None, ('air',)),
 'Carbon

It is also possible to use "filters" to narrow searches, e.g.

In [62]:
bio.search("carbon dioxide, fossil", filter={"categories": "urban", "name": "fossil"})

Excluding 7 filtered results


['Carbon dioxide, fossil' (kilogram, None, ('air', 'urban air close to ground')),
 'Carbon dioxide, fossil' (kilogram, None, ('air', 'non-urban air or from high stacks')),
 'Carbon dioxide, non-fossil' (kilogram, None, ('air', 'urban air close to ground')),
 'Carbon dioxide, non-fossil' (kilogram, None, ('air', 'non-urban air or from high stacks'))]

The database object is also iterable, allowing "home-made" searches through list comprehensions. This approach is better because one can add as many criteria as wanted:

In [64]:
[
    flow for flow in bio
    if "Carbon dioxide" in flow["name"]
    and ", fossil" in flow["name"]
    and flow["categories"] == ('air', 'lower stratosphere + upper troposphere')
]

['Carbon dioxide, fossil' (kilogram, None, ('air', 'lower stratosphere + upper troposphere'))]

Activities returned by searches or list comprehensions can be assigned to variables, but to do so, one needs to identify the activity by index. Based on the above, I can refine my filters to ensure the list comprehension only returns one activity, and then choose it without fear of choosing the wrong one..

In [65]:
co2 = [
    flow for flow in bio
    if "Carbon dioxide" in flow["name"]
    and ", fossil" in flow["name"]
    and flow["categories"] == ('air', 'lower stratosphere + upper troposphere')
][0]

In [66]:
co2

'Carbon dioxide, fossil' (kilogram, None, ('air', 'lower stratosphere + upper troposphere'))

**Exercise**: look for and assign to a variable an emission of nitrous oxide emitted to air in the "urban air" subcompartment.

In [67]:
co2.as_dict()

{'categories': ('air', 'lower stratosphere + upper troposphere'),
 'code': '16eeda8a-1ea2-408e-ab37-2648495058dd',
 'CAS number': '000124-38-9',
 'name': 'Carbon dioxide, fossil',
 'database': 'biosphere3',
 'unit': 'kilogram',
 'type': 'emission'}

In [68]:
len(bio)

4709

In [69]:
set(list(f["categories"] for f in bio))

{('air',),
 ('air', 'low population density, long-term'),
 ('air', 'lower stratosphere + upper troposphere'),
 ('air', 'non-urban air or from high stacks'),
 ('air', 'urban air close to ground'),
 ('economic', 'primary production factor'),
 ('inventory indicator', 'output flow'),
 ('inventory indicator', 'resource use'),
 ('inventory indicator', 'waste'),
 ('natural resource', 'biotic'),
 ('natural resource', 'fossil well'),
 ('natural resource', 'in air'),
 ('natural resource', 'in ground'),
 ('natural resource', 'in water'),
 ('natural resource', 'land'),
 ('soil',),
 ('soil', 'agricultural'),
 ('soil', 'forestry'),
 ('soil', 'industrial'),
 ('water',),
 ('water', 'fossil well'),
 ('water', 'ground-'),
 ('water', 'ground-, long-term'),
 ('water', 'ocean'),
 ('water', 'surface water')}

Let's leave the biosphere database here for now.

##### Methods

bw2_setup() also installed LCIA methods.

In [70]:
bw.methods

Methods dictionary with 762 objects, including:
	('CML v4.8 2016', 'acidification', 'acidification (incl. fate, average Europe total, A&B)')
	('CML v4.8 2016', 'climate change', 'global warming potential (GWP100)')
	('CML v4.8 2016', 'ecotoxicity: freshwater', 'freshwater aquatic ecotoxicity (FAETP inf)')
	('CML v4.8 2016', 'ecotoxicity: marine', 'marine aquatic ecotoxicity (MAETP inf)')
	('CML v4.8 2016', 'ecotoxicity: terrestrial', 'terrestrial ecotoxicity (TETP inf)')
	('CML v4.8 2016', 'energy resources: non-renewable', 'abiotic depletion potential (ADP): fossil fuels')
	('CML v4.8 2016', 'eutrophication', 'eutrophication (fate not incl.)')
	('CML v4.8 2016', 'human toxicity', 'human toxicity (HTP inf)')
	('CML v4.8 2016', 'material resources: metals/minerals', 'abiotic depletion potential (ADP): elements (ultimate reserves)')
	('CML v4.8 2016', 'ozone depletion', 'ozone layer depletion (ODP steady state)')
Use `list(this object)` to get the complete list.

One can load a random method:

In [71]:
bw.methods.random()

('ReCiPe 2016 v1.03, endpoint (E)',
 'ecosystem quality',
 'ecotoxicity: freshwater')

In [72]:
type(bw.methods.random())

tuple

In [73]:
method = bw.Method(('ReCiPe 2016 v1.03, endpoint (E)', 'ecosystem quality', 'ecotoxicity: freshwater'))

In [74]:
method

Brightway2 Method: ReCiPe 2016 v1.03, endpoint (E): ecosystem quality: ecotoxicity: freshwater

Here, the random method returns the tuple by which the method is identified. To get to an actual method, the following syntax is used:

Of course, a random method is probably not useful except to play around. To find an actual method, one can again use list comprehensions. Let's say I am interested in using the IPCC2013 100 years method:

In [107]:
bw.databases

Databases dictionary with 1 object(s):
	biosphere3

In [109]:
type(bw.Database("biosphere3"))

bw2data.backends.peewee.database.SQLiteBackend

In [112]:
type(list(bw.methods)[0])

tuple

In [113]:
type(bw.Method(('CML v4.8 2016', 'acidification', 'acidification (incl. fate, average Europe total, A&B)')))

bw2data.method.Method

In [104]:
[
    m for m in bw.methods
    if "ipcc" not in m[0].lower()
    and "2021" in m[0]
    and "100" in m[2]
    and "biogenic" not in m[2]
]

[('Ecological Scarcity 2021 no LT',
  'climate change no LT',
  'global warming potential (GWP100) no LT'),
 ('Ecological Scarcity 2021',
  'climate change',
  'global warming potential (GWP100)')]

I am interested in the second of these, and will assign it to a variable. I can will refine my search until there is one element in my list and then choose it by subscripting.

In [84]:
ipcc = bw.Method(('IPCC 2021',
  'climate change',
  'global warming potential (GWP100)'))
ipcc

Brightway2 Method: IPCC 2021: climate change: global warming potential (GWP100)

In [82]:
ipcc.metadata

{'description': '',
 'filename': 'LCIA_Implementation_3.9.xlsx',
 'unit': 'kg CO2-Eq',
 'abbreviation': 'ipcc-2021cg.53430e09e305748933330bd00b612547',
 'num_cfs': 5}

In [90]:
ipcc.load()

[(('biosphere3', '892f10da-7bf2-42b3-b171-221989661a05'), 0.006),
 (('biosphere3', 'f9749677-9c9f-4678-ab55-c607dfdc2cb9'), 1),
 (('biosphere3', '4cadbc9f-0472-4bb9-a942-df5b90e37878'), 20.6),
 (('biosphere3', '9ea11aea-7cb7-4b38-b936-a42c2a4bac97'), 0.437),
 (('biosphere3', 'fa0c2bee-8dd9-4f8a-8489-b1f3b43de958'), 1530),
 (('biosphere3', '818cee9e-231c-4b53-8ed2-47a0001802d5'), 161),
 (('biosphere3', 'f43b2b80-f5a6-4d31-9980-0754305a978d'), 5810),
 (('biosphere3', '033d3a16-e1a2-4ce8-8769-f4ea867801ca'), 860),
 (('biosphere3', '393d0862-04cc-45ae-8d08-9dc2a4461b35'), 164),
 (('biosphere3', '468bebea-17ee-49d3-b11d-c603d91c67c8'), 1.3),
 (('biosphere3', '400be66e-4383-482a-93e3-c005e96de261'), 2300),
 (('biosphere3', 'db907ff7-6cf0-4f88-9376-116262dfad6d'), 90.4),
 (('biosphere3', '27f61651-96a5-45c7-9873-832626cf6905'), 597),
 (('biosphere3', '15e54d5b-aba2-40a6-9aa8-aa2b8c432129'), 9600),
 (('biosphere3', 'df5dd437-2e12-4af6-8f7a-9c8224857dc5'), 12400),
 (('biosphere3', '5d1701c5-c4e

In [92]:
bw.get_activity(('biosphere3', '20185046-64bb-4c09-a8e7-e8a9e144ca98')).as_dict()

{'categories': ('air',),
 'code': '20185046-64bb-4c09-a8e7-e8a9e144ca98',
 'CAS number': '010024-97-2',
 'name': 'Dinitrogen monoxide',
 'database': 'biosphere3',
 'unit': 'kilogram',
 'type': 'emission'}

Of course, if I know exactly the method I want, and I know the syntax, I can simply type it out: `('IPCC 2021', 'climate change', 'GWP 100a')`

In [95]:
GWP_dict = {bw.get_activity(ef[0])["name"]: ef[1] for ef in ipcc.load()}
GWP_dict

{'Butane': 0.006,
 'Carbon dioxide, fossil': 1,
 'Chloroform': 20.6,
 'Ethane': 0.437,
 'Ethane, 1,1,1,2-tetrafluoro-, HFC-134a': 1530,
 'Ethane, 1,1,1-trichloro-, HCFC-140': 161,
 'Ethane, 1,1,1-trifluoro-, HFC-143a': 5810,
 'Ethane, 1,1-dichloro-1-fluoro-, HCFC-141b': 860,
 'Ethane, 1,1-difluoro-, HFC-152a': 164,
 'Ethane, 1,2-dichloro-': 1.3,
 'Ethane, 1-chloro-1,1-difluoro-, HCFC-142b': 2300,
 'Ethane, 2,2-dichloro-1,1,1-trifluoro-, HCFC-123': 90.4,
 'Ethane, 2-chloro-1,1,1,2-tetrafluoro-, HCFC-124': 597,
 'Ethane, chloropentafluoro-, CFC-115': 9600,
 'Ethane, hexafluoro-, HFC-116': 12400,
 'Ethane, pentafluoro-, HFC-125': 3740,
 'Tetrachloroethylene': 6.34,
 'Trichloroethylene': 0.044,
 'Methane, bromo-, Halon 1001': 2.43,
 'Methane, bromochlorodifluoro-, Halon 1211': 1930,
 'Methane, bromotrifluoro-, Halon 1301': 7200,
 'Methane, chlorodifluoro-, HCFC-22': 1960,
 'Methane, chloro-fluoro-, HCFC-31': 79.4,
 'Methane, chlorotrifluoro-, CFC-13': 16200,
 'Methane, dichloro-, HCC-30': 

In [103]:
pd.DataFrame.from_dict(GWP_dict, orient="index", columns=["factor"])

factor
Butane                                                  0.006
Carbon dioxide, fossil                                  1.000
Chloroform                                             20.600
Ethane                                                  0.437
Ethane, 1,1,1,2-tetrafluoro-, HFC-134a               1530.000
Ethane, 1,1,1-trichloro-, HCFC-140                    161.000
Ethane, 1,1,1-trifluoro-, HFC-143a                   5810.000
Ethane, 1,1-dichloro-1-fluoro-, HCFC-141b             860.000
Ethane, 1,1-difluoro-, HFC-152a                       164.000
Ethane, 1,2-dichloro-                                   1.300
Ethane, 1-chloro-1,1-difluoro-, HCFC-142b            2300.000
Ethane, 2,2-dichloro-1,1,1-trifluoro-, HCFC-123        90.400
Ethane, 2-chloro-1,1,1,2-tetrafluoro-, HCFC-124       597.000
Ethane, chloropentafluoro-, CFC-115                  9600.000
Ethane, hexafluoro-, HFC-116                        12400.000
Ethane, pentafluoro-, HFC-125                        3740.000
Tetrachloroethylene                                     6.340
Trichloroethylene                                       0.044
Methane, bromo-, Halon 1001                             2.430
Methane, bromochlorodifluoro-, Halon 1211            1930.000
Methane, bromotrifluoro-, Halon 1301                 7200.000
Methane, chlorodifluoro-, HCFC-22                    1960.000
Methane, chloro-fluoro-, HCFC-31                       79.400
Methane, chlorotrifluoro-, CFC-13                   16200.000
Methane, dichloro-, HCC-30                             11.200
Methane, dichlorodifluoro-, CFC-12                  12500.000
Methane, dichlorofluoro-, HCFC-21                     160.000
Methane, difluoro-, HFC-32                            771.000
Methane, fossil                                        29.800
Methane, monochloro-, R-40                              5.540
Methane, non-fossil                                    27.000
Methane, tetrachloro-, R-10                          2200.000
Methane, tetrafluoro-, R-14                          7380.000
Methane, trichlorofluoro-, CFC-11                    6230.000
Methane, trifluoro-, HFC-23                         14600.000
Monochloroethane                                        0.481
Nitrogen fluoride                                   17400.000
Propane                                                 0.020
Sulfur hexafluoride                                 25200.000
Ethane, 1,1,2-trichloro-1,2,2-trifluoro-, CFC-113    6520.000
Perfluoropentane                                     9220.000
Methane, from soil or biomass stock                    29.800
Carbon dioxide, from soil or biomass stock              1.000
Carbon dioxide, to soil or biomass stock               -1.000
Bromopropane                                            0.052
Ethane, 1,2-dichloro-1,1,2,2-tetrafluoro-, CFC-114   9430.000
Dinitrogen monoxide                                   273.000

Again, there are a bunch of methods associated with a method object. You can access these by typing ipcc_2013_method. and clicking tab.  
For example, metadata:

Question: where is this metadata?

Let's use the `load` method to see what is in the object:

This contains tuples with (elementary flow, characterization factors).

##### Nice display of data in methods 

**Exercise:** Create a dictionary with keys = elementary flow names and values = characterization factors for the `IPCC GWP100` method (including long-term emissions).

Bonus (optional): Generate a Pandas Series with the resulting dictionary. 

Enough said for now about methods.

#### 1.4) LCI databases

Outside of the classes, you would probably use the following code to "import" ecoinvent into your brightway installation:

On a "recent" computer, this can take a few minutes to be finished.

In [114]:
fp = r"/Users/romain/Documents/ecoinvent 3.9.1_cutoff_ecoSpold02/datasets"

ei39 = bw.SingleOutputEcospold2Importer(fp, "ecoinvent 3.9.1 cutoff")
ei39.apply_strategies()
ei39.statistics()

Extracting XML data from 21238 datasets
Extracted 21238 datasets in 20.35 seconds
Applying strategy: normalize_units
Applying strategy: update_ecoinvent_locations
Applying strategy: remove_zero_amount_coproducts
Applying strategy: remove_zero_amount_inputs_with_no_activity
Applying strategy: remove_unnamed_parameters
Applying strategy: es2_assign_only_product_with_amount_as_reference_product
Applying strategy: assign_single_product_as_activity
Applying strategy: create_composite_code
Applying strategy: drop_unspecified_subcategories
Applying strategy: fix_ecoinvent_flows_pre35
Applying strategy: drop_temporary_outdated_biosphere_flows
Applying strategy: link_biosphere_by_flow_uuid
Applying strategy: link_internal_technosphere_by_composite_code
Applying strategy: delete_exchanges_missing_activity
Applying strategy: delete_ghost_exchanges
Applying strategy: remove_uncertainty_from_negative_loss_exchanges
Applying strategy: fix_unreasonably_high_lognormal_uncertainties
Applying strategy: 

(21238, 674593, 0)

In [115]:
ei39.write_database()

Writing activities to SQLite3 database:
0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:15


Title: Writing activities to SQLite3 database:
  Started: 03/19/2024 11:06:10
  Finished: 03/19/2024 11:06:25
  Total time elapsed: 00:00:15
  CPU %: 98.80
  Memory %: 19.72
Created database: ecoinvent 3.9.1 cutoff


Brightway2 SQLiteBackend: ecoinvent 3.9.1 cutoff

Other code to import LCI databases in other formats are found [here](https://github.com/brightway-lca/brightway2-io/tree/master/bw2io/importers). We'll explore this later.

*We created* a new project, so let's make sure we set it as the "current" project.

In [116]:
bw.databases

Databases dictionary with 2 object(s):
	biosphere3
	ecoinvent 3.9.1 cutoff

To access the actual database, you need to use the Database method: 

Let's assign the database to a variable and see what we can do:

In [117]:
ei = bw.Database("ecoinvent 3.9.1 cutoff")
ei

Brightway2 SQLiteBackend: ecoinvent 3.9.1 cutoff

In [130]:
len(ei)

21238

Again, we can get an idea of useful methods and attributes by typing eidb. and Tab. Do this now.

In [ ]:
ei.

##### LCI activities

In the context of LCI databases, activities are the nodes "within the technosphere". They are therefore the columns in the technosphere matrix $A$.  
There are different ways to get access to an activity. Let's use the `random()` method for now to explore a random activity in the ecoinvent database.

In [129]:
act = ei.random()
act

'hardwood forestry, eucalyptus ssp., planted forest management' (kilogram, BR-MG, None)

In [128]:
act

'market for waste concrete, not reinforced' (kilogram, CH, None)

In [134]:
list(act.keys())

['comment',
 'classifications',
 'activity type',
 'activity',
 'database',
 'filename',
 'location',
 'name',
 'synonyms',
 'parameters',
 'authors',
 'type',
 'reference product',
 'flow',
 'unit',
 'production amount',
 'code']

To see what is stored in an activity object, let's convert our random act in a dictionary: 

Notice one important thing: **no exchanges**!  
Indeed, the exchanges and the activities are stored in two different tables of the `databases.db` database.  
It is possible, however, to iterate through the exchanges of the activities.

##### Searching and getting LCI activities

Searching and getting LCI activities is done exactly the same way as for activities in the biosphere3 database:

In [131]:
ei.search("electricity production")

['electricity production, oil' (kilowatt hour, CO, None),
 'electricity production, oil' (kilowatt hour, IN-UT, None),
 'electricity production, oil' (kilowatt hour, IN-WB, None),
 'electricity production, oil' (kilowatt hour, IN-JK, None),
 'electricity production, oil' (kilowatt hour, IN-GJ, None),
 'electricity production, oil' (kilowatt hour, IN-MZ, None),
 'electricity production, oil' (kilowatt hour, IN-NL, None),
 'electricity production, oil' (kilowatt hour, IN-AS, None),
 'electricity production, oil' (kilowatt hour, IN-MP, None),
 'electricity production, oil' (kilowatt hour, IN-OR, None),
 'electricity production, oil' (kilowatt hour, IN-MN, None),
 'electricity production, oil' (kilowatt hour, IN-TN, None),
 'electricity production, oil' (kilowatt hour, IN-HP, None),
 'electricity production, oil' (kilowatt hour, IN-SK, None),
 'electricity production, oil' (kilowatt hour, IN-RJ, None),
 'electricity production, oil' (kilowatt hour, IN-MH, None),
 'electricity production, o

**Exercise:** Return an activity for electricity production, coal-fired power plants in Germany

In [136]:
act = [
    a for a in ei
    if "electricity production" in a["name"]
    and "coal" in a["name"]
    and a["location"] == "DE"
][0]
act

'electricity production, hard coal' (kilowatt hour, DE, None)

In [137]:
act.as_dict()

{'comment': 'This dataset represents the production of high voltage electricity in an average hard coal power plant in Germany in 2012. "Hard coal" includes anthracite, coking coal and other bituminous coal according to the definition of the IEA electricity information 2014.\nAll exchanges in this dataset except for emissions of SO2, NOx and particles are production-weighted averages of the figures in the datasets in ecoinvent version two, which are valid for the following 14 countries: AT, BE, ES, FR, IT, NL, PT, DE, CZ, HR, PL, SK, CN, US. The exception values are production-weighted averages of version 2 countries and the new countries in version 3 (AU, CA, CL, ID, IN, KR, MX, MY, PE, RU, TH, TR, TW, TZ, UA, ZA)\nComments on the inventory data of emissions: \n- For the assessment of main characteristics (sulphur and ash content of coal, efficiency of the plant) and CO2 a bottom-up approach has been used. It consists in the collection of information about single plants.\n- The size d

In [138]:
act.exchanges()

('ecoinvent 3.9.1 cutoff', 'f30c46c44c2292cee46f9af3c1b8aca5')

#### LCI exchanges

**`Exchanges`** are the edges between nodes.

These can be:  
  - an edge between two activities within the technosphere (an element $a_{ij}$ of matrix $A$)  
  - edges between an activity in the technosphere and an activity in the "biosphere" (an element $b_{kj}$ of the biosphere matrix $B$).

One can iterate through **all** exchanges that have a given activity as `output`

In [142]:
for e in act.exchanges():
    if e["type"] == "technosphere":
        print(e["name"], e["amount"], e.get("location"), e["type"])

NOx retained, by selective catalytic reduction 0.00225844504021448 None technosphere
SOx retained, in hard coal flue gas desulfurisation 0.00594530831099196 None technosphere
chlorine, liquid 9.6514745308311e-05 None technosphere
hard coal 0.402466487935657 None technosphere
hard coal ash -0.00253833780160858 None technosphere
hard coal power plant 1.25469168900804e-11 None technosphere
light fuel oil 0.000164075067024129 None technosphere
residue from cooling tower -4.82573726541555e-05 None technosphere
water, completely softened 0.0579088471849866 None technosphere
water, decarbonised 1.44772117962467 None technosphere


In [144]:
1/1.25469168900804e-11/1e6

79700.85470085489

One can also iterate through subsets of the exchanges:  
  - Technosphere exchanges: exchanges linking to other activities in the technosphere, `activity.technosphere()`  
  - Biosphere exchanges: AKA elementary flows, linking to activities in the biosphere database `activity.biosphere()`  
  - Production exchange: the reference flow of the activity `activity.production`  

Let's assign a **technosphere exchange** to a variable to learn more about it:

Again, the type is a proxy (refer to the diagram above about the different translation layers).

Let's now look at a production exchange

**Exercise:** Assign a biosphere flow to a variable, and check the following:  
  - Is the output the same as for the technosphere exchange?  
  - From what database does the biosphere exchange come from?  
  - What is the amount of the exchange (i.e. the weight of the edge connecting the two activities)?
  
NOTE: If you get a ` list index out of range` error when trying to subscript your list comprehension, it means your list comprehension is empty, i.e. that there are no biosphere flows associated with the activity.

#### Loaded LCI databases

It is possible to load the entire database into a dictionary.  
This greatly speeds up work if you need to iterate over all activities or exchanges. The resulting object is quite big, so you should do this only if the gain in efficiency is worth it.

## 2) My first LCA

Brightway has a so-called LCA object.  
It is instantiated using `LCA(args)`.  
The only required argument is a functional unit, described by a dictionary with keys = activities and values = amounts (more [here](https://docs.brightwaylca.org/lca.html#specifying-a-functional-unit)).  
A second argument that is often passed is an LCIA method, passed using the method tuple.  

### 2.1) General syntax of LCA calculations

Let's create our first LCA object using our random activity and our IPCC method.  

In [145]:
ipcc

Brightway2 Method: IPCC 2021: climate change: global warming potential (GWP100)

In [146]:
act

'electricity production, hard coal' (kilowatt hour, DE, None)

In [147]:
ipcc.name

('IPCC 2021', 'climate change', 'global warming potential (GWP100)')

In [148]:
lca = bw.LCA({act: 1}, method=ipcc.name)

The steps to get to the impact score are as follows:

Let's not take a closer look at the LCA object and its methods/attributes. We'll do this by creating a new LCA object: 

In [150]:
lca.lci()

/opt/homebrew/Caskroom/miniforge/base/envs/bw/lib/python3.11/site-packages/scikits/umfpack/umfpack.py:736: UmfpackWarning: (almost) singular matrix! (estimated cond. number: 3.58e+13)
  warnings.warn(msg, UmfpackWarning)


In [176]:
lca.lcia()

In [177]:
lca.score

1.0545383927068062

In [170]:
act_dict, _, _ = lca.reverse_dict()

In [171]:
act_dict

{0: ('ecoinvent 3.9.1 cutoff', 'f32ba57ebce8c08759275f740a88f370'),
 1: ('ecoinvent 3.9.1 cutoff', '73a4483e483a28e471ceffb1e6ef8b0d'),
 2: ('ecoinvent 3.9.1 cutoff', 'fb126ed85de69f5cac7f991f022d9b5c'),
 3: ('ecoinvent 3.9.1 cutoff', '10302c31d0d4a1816bd8bb0032c0198f'),
 4: ('ecoinvent 3.9.1 cutoff', 'bad56a905eb6f8bed4e273051a455157'),
 5: ('ecoinvent 3.9.1 cutoff', '0d759a754439881bf0f5e934078dcedf'),
 6: ('ecoinvent 3.9.1 cutoff', '17c9670f75f1b10a3ea8850c4bffdccb'),
 7: ('ecoinvent 3.9.1 cutoff', '484ed4ed4737fac9849242b16d5a23c7'),
 8: ('ecoinvent 3.9.1 cutoff', '6ea399ff612d73f5a4cde92a17992053'),
 9: ('ecoinvent 3.9.1 cutoff', '2840342713a23d57067cb420745eb251'),
 10: ('ecoinvent 3.9.1 cutoff', '6caa34be583700c562617d1cc473c1b5'),
 11: ('ecoinvent 3.9.1 cutoff', '7d2e14ef80a94cebc66c3cff7c989496'),
 12: ('ecoinvent 3.9.1 cutoff', '362c5d4e230434904ed2821d5ac1098c'),
 13: ('ecoinvent 3.9.1 cutoff', 'ba4b47e0d1f15aff29e8b9ab1e1bb5ba'),
 14: ('ecoinvent 3.9.1 cutoff', 'fd558441fc3

In [182]:
for x, f in enumerate(lca.biosphere_matrix.sum(axis=1)):
    print(x, f[0])

0 [[0.01006901]]
1 [[0.00027231]]
2 [[0.00020911]]
3 [[0.00272545]]
4 [[0.00162764]]
5 [[0.00024168]]
6 [[0.00341045]]
7 [[0.0033492]]
8 [[0.00550001]]
9 [[0.49204292]]
10 [[0.0028502]]
11 [[0.02148005]]
12 [[4.10219654e-05]]
13 [[1.16570998e-11]]
14 [[0.00094576]]
15 [[0.05246912]]
16 [[7.78473343e-06]]
17 [[0.04962279]]
18 [[0.00139536]]
19 [[0.30004903]]
20 [[0.00380236]]
21 [[1.84414826e-05]]
22 [[0.00176296]]
23 [[0.00161923]]
24 [[0.00101908]]
25 [[0.0044262]]
26 [[6.21483786e-05]]
27 [[0.02382166]]
28 [[2.92230628e-05]]
29 [[4.50408273e-07]]
30 [[8.34426505]]
31 [[0.0024303]]
32 [[0.50531978]]
33 [[0.00103158]]
34 [[0.00836183]]
35 [[0.00018314]]
36 [[2.47882603e-05]]
37 [[0.278078]]
38 [[0.00016941]]
39 [[4.24165367e-07]]
40 [[0.00060194]]
41 [[0.00675674]]
42 [[9.22922218e-07]]
43 [[0.00313704]]
44 [[0.10897468]]
45 [[2.21063557]]
46 [[1.89599776e-07]]
47 [[3.74573967e-06]]
48 [[0.00347742]]
49 [[0.00635573]]
50 [[5.29866722e-06]]
51 [[0.24074345]]
52 [[0.00797644]]
53 [[2.155

In [175]:
for x in list(np.argwhere(lca.in)):
    print(bw.get_activity(act_dict[x[0]])["name"])

market for cyanuric chloride
market for silicon, multi-Si, casted
treatment of waste plastic, industrial electronics, municipal incineration
market for compressed air, 700 kPa gauge
electricity production, photovoltaic, 570kWp open ground installation, multi-Si
photovoltaic laminate production, a-Si
electricity voltage transformation from medium to low voltage
market for used locomotive
market for bottom ash, MSWI-WWT, WW from medium density board production
market for uranium hexafluoride
maintenance, goods wagon
market for sodium silicate, solid
market for graphite
market for gravel, crushed
primary lead production from concentrate
electricity voltage transformation from medium to low voltage
market for aniline
market for inorganic potassium fertiliser, as K2O
calcareous marl production
treatment of waste paper, unsorted, sorting
treatment of waste graphical paper, unsanitary landfill, dry infiltration class (100mm)
planing, beam, softwood, u=20%
nutrient supply from urea ammonium ni

### 2.2) the `demand` attribute

In [155]:
lca.biosphere_matrix

<2420x21238 sparse matrix of type '<class 'numpy.float64'>'
	with 407351 stored elements in Compressed Sparse Row format>

To access the actual activity from the demand, you would do this:

There are also other attributes that have simply not been built yet, such as the `demand_array` and the `score`. To generate them, we first need to actually build the matrices. This will be done when calling the `.lci()` method.

### 2.3) Reminder of the system that needs to be solved in calculating an LCI

Before actually running the `.lci()` method, here's a quick refresher of the actual calculation that Brightway will need to do to calculate the inventory:  

$g=BA^{-1}f$  

where:  

  - $A$ = the technosphere matrix  
  - $B$ = the biosphere matrix (matrix with elementary flows)  
  - $f$ = the final demand vector  
  - $g$ = the inventory  

**Discussion:** Knowing what you do about the structure of Brightway (notably, activities and exchanges), what needs to happen to generate these matrices?

To consider:  
  - how should the order of the rows and columns be determined?  
  - how should we keep track of what is in each row and column?  
  - The parameters in the matrices are sometimes actually probability distribution functions - how should we consider this uncertainty information?  
  - The matrices are *sparse*, i.e. they are mostly made up of zeros. Should we consider this? Why? How?

### 2.4) Building the matrices

#### Structured arrays

LCI data imported in Brightway is stored in the `databases.db` database, discussed above.  
It is also stored in [numpy *structured arrays*](https://docs.scipy.org/doc/numpy/user/basics.rec.html). 

**Exercise**(not core): Load the structured array of the ecoinvent database you are working with now.

In this array:  
  - `input` and `output` columns are integers that map to an activity. This mapping is found in the `mapping.pickle` file in the `project` directory and it looks something like this:

  - `row` and `col`store *dummy* placeholder information about the location of the parameter in the matrices. 
  - the `type` indicates whether the exchange is a reference flow (`type`=0), technosphere exchange (`type`=1) or elementary flow (`type`=2).  
  - the other columns deal with uncertainty data. We'll cover that later, but one can always read about these columns in the [`stats_arrays` documentation](http://stats-arrays.readthedocs.io/en/latest/)

When the `.lci()` method is called, the structured arrayas are used to build matrices. The code responsible to do this is in the [`MatrixBuilder` class methods](https://2.docs.brightway.dev/technical/bw2calc.html#matrix-builder). 

The method `MatrixBuilder.build_dictionary` is used to take input and output values, respectively, and figure out which rows and columns they correspond to. The actual code is succinct - only one line - but what it does is:
  - Get all unique values, as each value will appear multiple times
  - Sort these values
  - Give them integer indices, starting with zero
This information on row and column indices is sufficient to build matrices. These matrices are build in a [COOrdinate sparse matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html) format, where, for each exchange, three values are required: row position, column position, and amount (the actual value). The sparse matrices are actually stored in [CSR format](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix), but this is a detail.  

Some more details are are found [here](https://2.docs.brightway.dev/technical/bw2calc.html#matrix-builders).  

Let's now finally run the `.lci()`.

Here's what the structured arrays *now* look like:  

We see that the row and col numbers are no longer dummy variables, but that they actually have real matrix indices.

#### Dictionaries that map between incides and activities

One of the useful things that the `MatrixBuilder` produces are `dictionaries` that map row and column numbers to the keys of activities.  There are three such dictionaries, all directly accessible as attributes of the LCA object:
 - `activity_dict`: Columns in the technosphere matrix $A$ or biosphere matrix $B$
 - `product_dict` : Rows in the technosphere matrix $A$  
 - `biosphere_dict`: Rows in the biosphere matrix $B$

In [162]:
list(lca.demand.keys())[0].key

('ecoinvent 3.9.1 cutoff', 'f30c46c44c2292cee46f9af3c1b8aca5')

In [163]:
lca.activity_dict[('ecoinvent 3.9.1 cutoff', 'f30c46c44c2292cee46f9af3c1b8aca5')]

18990

Here what this dictionary looks like:

In [ ]:
actitivty_dict, product_dict, biosphere_dict = lca.

So, if I know the key to my activity (which, again,  is a `tuple` consisting of the database name and the activity code), I can read the column index (from `activity_dict`) or row index (from `product_dict` or `biosphere_dict` for the $A$ or $B$ matrices, respectively). 

Let's find out what column is associated with the activity that is producing our final demand as reference flow.

While this is useful, it is often more useful to determine what a row or column in the matrices actually refers to. In these cases, we need a dictionary that maps row or column indices to activity keys, and not the opposite.  
We can do this by reversing our dictionaries:

As a convenience, Brightway offers a method that will generate the three reverse dictionaries simultaneously.  
`.reverse_dict()` returns three reverse dicts (reverse activity dict,  reverse product dict,  reverse biosphere dict) *in that order*. The syntax for creating and assigning these reverse dicts is therefore: 

#### $A$ and $B$ Matrices

We can also access the matrices that were constructed. Let's look at the technosphere matrix ($A$).  
The ** $A$ matrix**, with each element $a_{ij}$ provides information on the amount of input or output of product $i$ from activity $j$. When $i=j$, the element $a_{ij}$ is the *reference flow* for the activity described in the column.

I am told that the dimensions of the matrix is $n*n$ where $n$ is the number of activities in my product system, and that the amount of actually stored elements is much less than $n^2$ (because the matrix is *sparse* and zero values are not stored).  

We can have an idea of what it stores by printing it out:

It therefore stores both the coordinates and the values (as expected).
We can slice this matrix using coordinates. For example, let's say we wanted a view of the exchanges associated with the unit process providing our functional unit.  
We already know found the column number for that activity: 

To return the whole column from the $A$ matrix, we therefore slice the $A$ matrix.  
Python notes:  
  - In Python, slicing is done using []
  - We specify rows first, then columns  
  - `:` refers to "the whole row" or "the whole column" (depending if it is passed first or second in the []) 

Printing this out gives:

Not too useful: it would be better to get the *names* to these exchanges.  
We need to do two things:  
  - Get the indices from the CSR matrix (we can do this by converting it to a sparse matrix in `COOrdinate` format first)  
  - Get the activity code for the each index (we can do this using the reverse of the `activity_dict`)  
  - Use `get_activity` to access the actual names of the activities.  

1) Converting the CSR matrix to a COO matrix:  

It is still a sparse matrix with the same number of elements, and it looks quite like the CSR version when we print it out:

However, we can directly access the rows and column indices using `row` and `col`:  

2) Get the activity code for each element using the **reverse** product dictionary we produced above:

It would be even nicer to get the names for these:

We can put these in a neat Pandas Series, with actual names and amounts:

Alternative way to generate similar information without even looking at the matrices:

Note the differences:  
  - The reference flow is not there (activity.technosphere() only returns technoshere exchanges where the input is not equal to the output)  
  - The values are positive, not negative (because the $A$ matrix is $I-Z$ where $Z$ contains the information on these inputs.

**Exercise**: Create a Pandas Series with the elementary flows of the activity supplying the reference flow for myFirstLCA.

#### Demand array $f$

The demand array is the $f$ in $As=f$. 
It is an attribute of the LCA object:

Looks like it is all zeros, but not so:

So where is the one? We can know this by using our `activity_dict`.

### 2.5) Solution to the inventory calculation

We saw above how `.lci()` produced the $A$ and $B$ matrices.  
`.lci()` also *solves* the equation $As=f$ and calculated the inventory by multiplying the solution to this equation by the biosphere matrix.  

#### Supply array

Vector containing the amount each activity will need to provide to meet the functional demand, i.e. $s=A^{-1}f$.

**Inventory matrix**  
Contains the inventory *by activity* (i.e. not summed). In other words, we do not have $g=BA^{-1}f$, but rather $G=B \cdot diag(A^{-1}f)$

We can aggregate the LCI results along the columns (i.e. calculate the cradle-to-gate inventory):

**Exercise:** Get the total (cradle-to-gate) emissions of nitrous oxide emitted to air in the "urban air" subcompartment.

### 2.7) LCIA

The LCIA calculation is done via the `.lcia()` method.

A number of other matrices are now available:

Question: would there be more, less or just as many elements in the inventory matrix as there are in the characterized inventory matrix?

The overall score is now an attribute of the LCA object: 

We also could have determined what this score was by summing the elements of our `characterized_inventory` matrix:

We could also have calculated it by multiplying the inventory and characterization factors ourselves:

We could also calculate the score by elementary flow (summing columns for each rows), irrespective of the unit process that produced it:

Notice that is has **two** dimensions. The result is in fact a one-dimensional matrix:

To convert it to an array (probably more useful for many purposes), you can use any of the following approaches:

**Exercise:** Create a Pandas series that has the scores per unit process, sorted by value (contribution analysis)

## 3) My second LCA: Comparative LCA

Let's choose two activities to compare, say Swiss electricity produced from respectively a run-of-river hydropower plant and a wind turbine.

**Exercise**: assign the two activities to variables `hydro` and `wind` respectively.

Let's also compare these according to their carbon footprint as measured with the IPCC method we already selected above:

#### One at a time approach:

**Exercise:** Do the LCA for `wind`:

Do one "delta" LCA:

## 4) My third LCA - Multiple impact categories

Say we want to evaluate the indicator results for our randomAct for all EF v3 categories (with long-term emissions).

Simplest way: for loop, using `switch method`

## Revising my second and third LCA with `MultiLCA`

The `MultiLCA` allows the calculation of LCA results for multiple functional units and impact categories.  
One simply needs to create a `calculation setup`, i.e. a named set of functional units and LCIA methods.

Calculation setups: dictionary with lists of functional units and methods.

You can also create "fuller" DataFrames. Here is with code from [here](http://stackoverflow.com/questions/42984831/create-a-dataframe-from-multilca-results-in-brightway2): 

## Uncertainty

Values of exchanges in datasets may come with uncertainty.
See [documentation](https://stats-arrays.readthedocs.io/en/latest/) of `stats_array`.

Let's find out how uncertainty is defined in ecoinvent:

### Running a stochastic analysis (Monte Carlo)

Let's visualize the results:

## Let's now look at the import and export of data

Open the second notebook.